File: backend/app/models/pipeline.py
"""Pipeline model — defines a scheduled data sync job."""
from sqlalchemy import Column, Integer, String, Text, Boolean, DateTime, ForeignKey, JSON
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship

from app.database import Base


class Pipeline(Base):
    """A scheduled job that syncs data from a source connector to a local SQLite table.

    schedule examples: "0 * * * *" (hourly), "0 0 * * *" (daily), "*/30 * * * *" (every 30 min)
    status: active | paused
    """

    __tablename__ = "pipelines"

    id = Column(Integer, primary_key=True, index=True)
    project_id = Column(Integer, ForeignKey("projects.id", ondelete="CASCADE"), nullable=False, index=True)
    name = Column(String(200), nullable=False)
    description = Column(Text)

    # Source: which datasource + object to pull from
    datasource_id = Column(String(100), nullable=False)   # references datasource id in YAML config
    object_type = Column(String(100), nullable=False)     # ontology object name
    filters = Column(JSON, default=[])                    # optional filter conditions

    # Target: local SQLite table name
    target_table = Column(String(100), nullable=False)

    # Schedule
    schedule = Column(String(50), nullable=False, default="0 * * * *")  # cron expression
    status = Column(String(20), nullable=False, default="active")        # active | paused

    # Tracking
    last_run_at = Column(DateTime(timezone=True))
    last_run_status = Column(String(20))   # success | error | running
    last_run_rows = Column(Integer)
    last_error = Column(Text)

    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())

    project = relationship("Project", back_populates="pipelines")

File: frontend/src/services/pipeline.ts
import api from './api';

export interface PipelineData {
  id: number;
  name: string;
  description: string | null;
  datasource_id: string;
  object_type: string;
  filters: any[];
  target_table: string;
  schedule: string;
  status: string;
  last_run_at: string | null;
  last_run_status: string | null;
  last_run_rows: number | null;
  last_error: string | null;
  created_at: string | null;
}

export interface PipelineCreate {
  name: string;
  description?: string;
  datasource_id: string;
  object_type: string;
  filters?: any[];
  target_table: string;
  schedule?: string;
}

export interface PipelineRunData {
  id: number;
  status: string;
  rows_synced: number | null;
  duration_seconds: number | null;
  error_message: string | null;
  triggered_by: string;
  created_at: string | null;
}

export const pipelineService = {
  list: async (projectId: number): Promise<PipelineData[]> => {
    const res = await api.get(`/projects/${projectId}/pipelines`);
    return res.data.pipelines;
  },

  create: async (projectId: number, data: PipelineCreate): Promise<PipelineData> => {
    const res = await api.post(`/projects/${projectId}/pipelines`, data);
    return res.data;
  },

  update: async (projectId: number, pipelineId: number, data: Partial<PipelineCreate & { status: string }>): Promise<PipelineData> => {
    const res = await api.put(`/projects/${projectId}/pipelines/${pipelineId}`, data);
    return res.data;
  },

  delete: async (projectId: number, pipelineId: number): Promise<void> => {
    await api.delete(`/projects/${projectId}/pipelines/${pipelineId}`);
  },

  run: async (projectId: number, pipelineId: number): Promise<{ success: boolean; rows?: number; error?: string }> => {
    const res = await api.post(`/projects/${projectId}/pipelines/${pipelineId}/run`);
    return res.data;
  },

  listRuns: async (projectId: number, pipelineId: number, limit = 20): Promise<PipelineRunData[]> => {
    const res = await api.get(`/projects/${projectId}/pipelines/${pipelineId}/runs`, { params: { limit } });
    return res.data.runs;
  },
};
